Tasks/AzureResourceManagerTemplateDeploymentV3/operations/DeploymentScopeBase.ts

```diff
@@ -37,10 +49,10 @@ export class DeploymentScopeBase {
         }
 
         this.deploymentParameters = params;
-        await this.performAzureDeployment(3);
+        await this.performAzureDeployment(3, await this.getServicePrincipalName());
     }
 
-    protected async performAzureDeployment(retryCount = 0): Promise<void> {
+    protected async performAzureDeployment(retryCount = 0, spnName: string): Promise<void> {
```

Comment:
This method is already async so why can't `await this.getServicePrincipalName()` be called here directly rather than passing as argument multiple places.

Reply:
2 reasons:
1. If we keep await inside callback, we'll need to make callback async and that won't work with our current implementation
2. If we're keeping it ouside callback in the function, API call is made more than once, with each retry

Reply:
we can have option #2 with a cached SPN name (using a private variable) so that it actually makes rest calls only once.

Reply:
I've updated